build_system/src/config.rs

```diff
@@ -159,30 +187,151 @@ impl ConfigInfo {
+            return Err(format!(
+                "{}: invalid commit hash `{}`",
+                commit_hash_file.display(),
+                commit,
+            ));
+        }
+        let output_dir = output_dir.join(commit);
+        if !output_dir.is_dir() {
+            std::fs::create_dir_all(&output_dir).map_err(|err| {
+                format!(
+                    "failed to create folder `{}`: {:?}",
+                    output_dir.display(),
+                    err,
+                )
+            })?;
+        }
+        let output_dir = output_dir.canonicalize().map_err(|err| {
+            format!(
+                "Failed to get absolute path of `{}`: {:?}",
+                output_dir.display(),
+                err
+            )
+        })?;
+
+        let libgccjit_so_name = "libgccjit.so";
+        let libgccjit_so = output_dir.join(libgccjit_so_name);
+        if !libgccjit_so.is_file() && !self.no_download {
+            // Download time!
+            let tempfile_name = "libgccjit.so.download";
+            let tempfile = output_dir.join(tempfile_name);
+            let is_in_ci = std::env::var("GITHUB_ACTIONS").is_ok();
+
+            let url = format!(
+                "https://github.com/antoyo/gcc/releases/download/master-{}/libgccjit.so",
+                commit,
+            );
+
+            println!("Downloading `{}`...", url);
+            // Try curl. If that fails and we are on windows, fallback to PowerShell.
+            let mut ret = run_command_with_output(
+                &[
+                    &"curl",
+                    &"--speed-time",
+                    &"30",
+                    &"--speed-limit",
+                    &"10", // timeout if speed is < 10 bytes/sec for > 30 seconds
+                    &"--connect-timeout",
+                    &"30", // timeout if cannot connect within 30 seconds
+                    &"-o",
+                    &tempfile_name,
+                    &"--retry",
+                    &"3",
+                    &"-SRfL",
+                    if is_in_ci { &"-s" } else { &"--progress-bar" },
+                    &url.as_str(),
+                ],
+                Some(&output_dir),
+            );
+            if ret.is_err() && cfg!(windows) {
+                eprintln!("Fallback to PowerShell");
+                ret = run_command_with_output(
+                    &[
+                        &"PowerShell.exe",
+                        &"/nologo",
+                        &"-Command",
+                        &"[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12;",
+                        &format!(
+                            "(New-Object System.Net.WebClient).DownloadFile('{}', '{}')",
+                            url,
+                            tempfile_name,
+                        ).as_str(),
+                    ],
+                    Some(&output_dir),
+                );
             }
```

Comment:
Please put this in a separate function (or two separate functions with `#[cfg(windows)]` if you don't want to run `curl` on Windows) since this is platform-specific.

Reply:
:+1: